prov/efa/src/efa.h

```diff
@@ -0,0 +1,410 @@
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef EFA_H
+#define EFA_H
+
+#include "config.h"
+#include "efa_compiler.h"
+
+#include <asm/types.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <poll.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <pthread.h>
+#include <sys/epoll.h>
+#include <uthash.h>
+
+#include "infiniband/efa_arch.h"
+#include "infiniband/efa_verbs.h"
+#include <rdma/fabric.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_errno.h>
+
+#include "ofi.h"
+#include "ofi_enosys.h"
+#include "ofi_list.h"
+#include "ofi_util.h"
+#include "ofi_file.h"
+
+#define EFA_PROV_NAME "efa"
+#define EFA_PROV_VERS FI_VERSION(3, 0)
+
+#define EFA_WARN(subsys, ...) FI_WARN(&efa_prov, subsys, __VA_ARGS__)
+#define EFA_TRACE(subsys, ...) FI_TRACE(&efa_prov, subsys, __VA_ARGS__)
+#define EFA_INFO(subsys, ...) FI_INFO(&efa_prov, subsys, __VA_ARGS__)
+#define EFA_INFO_ERRNO(subsys, fn, errno) \
+	EFA_INFO(subsys, fn ": %s(%d)\n", strerror(errno), errno)
+#define EFA_WARN_ERRNO(subsys, fn, errno) \
+	EFA_WARN(subsys, fn ": %s(%d)\n", strerror(errno), errno)
+#define EFA_DBG(subsys, ...) FI_DBG(&efa_prov, subsys, __VA_ARGS__)
+
+#define EFA_ABI_VER_MAX_LEN 8
+
+#define EFA_WCE_CNT 1024
+
+#define EFA_EP_TYPE_IS_RDM(_info) \
+	(_info && _info->ep_attr && (_info->ep_attr->type == FI_EP_RDM))
+
+#define EFA_MEM_ALIGNMENT (64)
+
+#define EFA_DEF_CQ_SIZE 1024
+#define EFA_MR_IOV_LIMIT 1
+#define EFA_MR_SUPPORTED_PERMISSIONS (FI_SEND | FI_RECV)
+
+#define EFA_DEF_NUM_MR_CACHE 36
+
+extern int efa_mr_cache_enable;
+extern size_t efa_mr_max_cached_count;
+extern size_t efa_mr_max_cached_size;
+extern int efa_mr_cache_merge_regions;
+
+extern struct fi_provider efa_prov;
+extern struct util_prov efa_util_prov;
+
+struct efa_fabric {
+	struct util_fabric	util_fabric;
+};
```

Comment:
Do you even need struct efa_fabric, since it's nothing more than a container for a single field?